libs/ml/include/ml/ops/loss_functions/mean_square_error.hpp

```diff
@@ -22,37 +22,66 @@
 #include <vector>
 
 #include "math/ml/loss_functions/mean_square_error.hpp"
-#include "ml/ops/loss_functions/criterion.hpp"
+#include "ml/ops/ops.hpp"
 
 namespace fetch {
 namespace ml {
 namespace ops {
 
 template <class T>
-class MeanSquareError : public Criterion<T>
+class MeanSquareError : public Ops<T>
 {
 public:
-  using ArrayType    = T;
-  using Datatype     = typename ArrayType::Type;
-  using ArrayPtrType = std::shared_ptr<ArrayType>;
+  using ArrayType     = T;
+  using DataType      = typename ArrayType::Type;
+  using SizeType      = typename ArrayType::SizeType;
+  using VecTensorType = typename Ops<T>::VecTensorType;
 
   MeanSquareError()          = default;
   virtual ~MeanSquareError() = default;
 
-  virtual typename ArrayType::Type Forward(std::vector<ArrayType> const &inputs)
+  virtual void Forward(VecTensorType const &inputs, ArrayType &output)
   {
     assert(inputs.size() == 2);
-    assert(inputs[0].shape() == inputs[1].shape());
+    assert(inputs.at(0).get().shape() == inputs.at(1).get().shape());
 
-    typename ArrayType::Type ret = fetch::math::MeanSquareError(inputs[0], inputs[1]);
-    return ret;
+    SizeType batch_dim = inputs.at(0).get().shape().size() - 1;
+    for (SizeType i{0}; i < inputs.at(0).get().shape().at(batch_dim); i++)
+    {
+      typename ArrayType::Type ret = fetch::math::MeanSquareError(
+          inputs.at(0).get().Slice(i, batch_dim), inputs.at(1).get().Slice(i, batch_dim));
+      output(0, i) = ret;
+    }
   }
 
-  virtual ArrayType Backward(std::vector<ArrayType> const &inputs)
+  // grad[0]=2*err*(in[0]-in[1])/mean_size, grad[1]=-2*err*(in[0]-in[1])/mean_size,
+  virtual std::vector<ArrayType> Backward(VecTensorType const &inputs,
+                                          ArrayType const &    error_signal)
   {
     assert(inputs.size() == 2);
-    assert(inputs[0].shape() == inputs[1].shape());
-    return fetch::math::Subtract(inputs[0], inputs[1]);
+    assert(inputs.at(0).get().shape() == inputs.at(1).get().shape());
+
+    ArrayType return_signal_1(inputs.front().get().shape());
```

Comment:
dont construct new tensors in the training loop if you don't have to. Make these private member variables; cache the input shape - if it hasn't changed, just fill the private member tensors - otherwise reshape them.